### ISSUE TITLE
Update to PKCS#11 v2.40

### ISSUE THREAD

**Issue by bellgrim** [opendnssec/SoftHSMv2#130]
Update SoftHSM to the latest PKCS#11 version. This requires new header files and updated source code.

Migrated from:
https://issues.opendnssec.org/browse/SOFTHSM-6

**Comment by bellgrim** [comment]
Francis Dupont added a comment - 2014-11-19 15:20
Still waiting for the v2.40 include files. Note the specs are ready (OASIS Committee Specifications).

**Comment by jariq** [comment]
There are several bugs in PKCS#11 v2.40 which prevent complete headers from being published so IMO this will have to wait until PKCS#11 v2.41 is published. See [discussion on pkcs11-comment mailing list](https://lists.oasis-open.org/archives/pkcs11-comment/201505/msg00002.html) for more details.

**Comment by bellgrim** [comment]
The v2.40 PKCS#11 headers are now available. URLs are
http://docs.oasis-open.org/pkcs11/pkcs11-base/v2.40/errata01/csprd01/include/pkcs11-v2.40/pkcs11.h
(and pkcs11[ft].h)

**Comment by jariq** [comment]
**No they are not**. Those are only draft headers published for public review of PKCS#11 Errata01 (currently unpublished bugfix release for v2.40) and they still contain some [serious problems](https://lists.oasis-open.org/archives/pkcs11-comment/201602/msg00001.html) that need to be resolved by PKCS#11 Technical Committee first. My professional suggestion: **Don't use them yet.**

**Comment by bellgrim** [comment]
Ok, did not read the details about the files. Just forwarded some information that was sent to me.

**Comment by jariq** [comment]
[PKCS#11 v2.40 Errata 01 docs have been approved and published](https://lists.oasis-open.org/archives/pkcs11-comment/201606/msg00002.html) along with [header files](http://docs.oasis-open.org/pkcs11/pkcs11-base/v2.40/errata01/os/include/pkcs11-v2.40/). However I am not sure whether header license is compatible with SoftHSM license:

```
/* Copyright (c) OASIS Open 2016. All Rights Reserved./
 * /Distributed under the terms of the OASIS IPR Policy,
 * [http://www.oasis-open.org/policies-guidelines/ipr], AS-IS, WITHOUT ANY
 * IMPLIED OR EXPRESS WARRANTY; there is no warranty of MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE or NONINFRINGEMENT of the rights of others.
 */
```

**Comment by bellgrim** [comment]
Fix in #287
